core/src/main/scala/com/salesforce/op/stages/impl/tuning/DataCutter.scala

```diff
@@ -75,40 +78,91 @@ class DataCutter(uid: String = UID[DataCutter]) extends Splitter(uid = uid) with
+
+    if (!isSet(labelsToKeep)) {
+      val labelCounts = data.groupBy(labelColName).count().persist()
+      val res = estimate(labelCounts)
+      labelCounts.unpersist()
+      setLabels(res.labelsKept.distinct, res.labelsDropped.distinct, res.labelsDroppedTotal)
+    }
+
+    val labelMetaArr = getLabelsFromMetadata(data)
+    val labelSet = getLabelsToKeep.toSet
 
-    import data.sparkSession.implicits._
+    log.info(s"Dropping rows with columns not in $labelSet")
 
-    val labels = data.map(r => r.getDouble(0) -> 1L)
-    val labelCounts = labels.groupBy(labels.columns(0)).sum(labels.columns(1)).persist()
-    val (resKeep, resDrop) = estimate(labelCounts)
-    labelCounts.unpersist()
-    setLabels(resKeep, resDrop)
+    // Update metadata that spark.ml Classifier is using tp determine the number of classes
+    //
+    val na = NominalAttribute.defaultAttr.withName(labelColName)
+    val metadataNA = if (labelMetaArr.isEmpty) {
+      log.info("setting num vals " + labelSet.max.toInt + 1)
+      na.withNumValues(labelSet.max.toInt + 1)
+    } else {
+      val newLabelMetaArr = labelMetaArr
+        .zipWithIndex
+        .collect {
+          case (label: String, idx: Int) if labelSet.contains(idx.toDouble) => label
+        }
+
+      na.withValues(newLabelMetaArr)
+    }
 
-    summary = Option(DataCutterSummary(labelsKept = getLabelsToKeep, labelsDropped = getLabelsToDrop))
-    summary
+    // filter low cardinality labels out of the dataframe to reduce the volume and  to keep
+    // it in sync with the new metadata.
+    //
+    val dataPrep = data
+      .filter(r => labelSet.contains(r.getDouble(labelColIdx)))
+      .withColumn(labelColName, data(labelColName).as("_", metadataNA.toMetadata))
+
+    summary = Option(DataCutterSummary(
+      labelsKept = getLabelsToKeep,
+      labelsDropped = getLabelsToDrop,
+      labelsDroppedTotal = getLabelsDroppedTotal
+    ))
+
+    if (isSet(cacheValidatedDFForTesting) && $(cacheValidatedDFForTesting)) {
+      cachedDataFrameForTesting = Option(dataPrep)
+    }
+
+    PrevalidationVal(summary, Option(dataPrep))
   }
 
-  /**
-   * Removes labels that should not be used in modeling
-   *
-   * @param data first column must be the label as a double
-   * @return Training set test set
-   */
-  override def validationPrepare(data: Dataset[Row]): Dataset[Row] = {
-    val dataPrep = super.validationPrepare(data)
 
-    val keep: Set[Double] = getLabelsToKeep.toSet
-    val dataUse = dataPrep.filter(r => keep.contains(r.getDouble(0)))
+  def getLabelsFromMetadata(data: DataFrame): Array[String] = {
+    val labelSF = data.schema.head
+    val labelColMetadata = labelSF.metadata
+    log.info(s"Raw label column metadata: $labelColMetadata")
 
-    dataUse
+    Try {
+      labelColMetadata
+        .getMetadata("ml_attr")
+        .getStringArray("vals")
+    }
+    .recover { case nonFatal =>
+      log.warn("Recovering non-fatal exception using num_vals", nonFatal)
```

Comment:
I dont understand these exceptions

Reply:
I wanted to avoid too elaborate of the logic for checking whether these attributes are set and set correctly. The most common case is NoSuchKey when the attributes are not there.

Reply:
it's not really a user friendly message either. please provide some more valuable information.